sigstore/_internal/tuf.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2022 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from pathlib import Path
+from typing import List, Optional, Tuple
+from urllib import parse
+
+import appdirs
+from tuf.ngclient import Updater
+
+from sigstore._utils import read_embedded
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_TUF_URL = "https://sigstore-tuf-root.storage.googleapis.com/"
+STAGING_TUF_URL = "https://tuf-root-staging.storage.googleapis.com/"
+
+# for tests to override
+_fetcher = None
+
+
+def _get_dirs(url: str) -> Tuple[Path, Path]:
+    """
+    Given a TUF repository URL, return suitable local metadata and cache directories.
+
+    These directories are not guaranteed to already exist.
+    """
+
+    builder = appdirs.AppDirs("sigstore-python", "sigstore")
+    tuf_base = parse.quote(url, safe="")
+
+    data_dir = Path(builder.user_data_dir)
+    cache_dir = Path(builder.user_cache_dir)
+
+    return (data_dir / tuf_base), (cache_dir / tuf_base)
+
+
+class TrustUpdater:
+    """Internal trust root (certificates and keys) downloader.
+
+    TrustUpdater discovers the currently valid certificates and keys and
+    securely downloads them from the remote TUF repository at 'url'.
+
+    TrustUpdater expects to find an initial root.json in either the local
+    metadata directory for this URL, or (as special case for the sigstore.dev
+    production and staging instances) in the application resources.
+    """
+
+    def __init__(self, url: str) -> None:
+        self._repo_url = url
+        self._updater: Optional[Updater] = None
+
+        self._metadata_dir, self._targets_dir = _get_dirs(url)
+
+        # intialize metadata dir
+        tuf_root = self._metadata_dir / "root.json"
+        if not tuf_root.exists():
+            if self._repo_url == DEFAULT_TUF_URL:
+                fname = "root.json"
+            elif self._repo_url == STAGING_TUF_URL:
+                fname = "staging-root.json"
+            else:
+                raise Exception(f"TUF root not found in {tuf_root}")
+
+            self._metadata_dir.mkdir(parents=True, exist_ok=True)
+            root_json = read_embedded(fname)
+            with tuf_root.open("wb") as io:
+                io.write(root_json)
+
+        # intialize targets cache dir
+        # TODO: Pre-populate with any targets we ship with sources
+        self._targets_dir.mkdir(parents=True, exist_ok=True)
+
+        logger.debug("TUF metadata: %s", self._metadata_dir)
```

Comment:
We've been using f-strings for this type of thing elsewhere in the codebase.

Reply:
Yeah, let's normalize on f-strings here. @tetsuo-cpp feel free to push to this branch!

Reply:
[d9aa72c](https://github.com/sigstore/sigstore-python/pull/351/commits/d9aa72c2e04477bf5e1d0b818f9ac31d4c91a651)